test_partner_client.py

```diff
@@ -0,0 +1,131 @@
+                datetime.strptime(date_str, '%Y%m%d')
+            except ValueError:
+                raise AssertionError(
+                    "Incorrect date format on {}, should be YYYYMMDD".format(date_str))
+
+        verify_date_format(arrival_date)
+        self.arrival_date = arrival_date
+
+        verify_date_format(result_date)
+        self.result_date = result_date
+
+        if len(comment) > 255:
+            raise AssertionError(
+                "comments cannot have more than 255 characters")
+
+        self.comment = comment
+
+        valid_covid_responses = ["negative", "positive", "failed"]
+
+        if cov19_result not in valid_covid_responses:
+            raise AssertionError("cov_19_result has to have one of the following values: {}".format(
+                ", ".join(valid_covid_responses)))
+
+    def get_as_dict(self):
+        return vars(self)
+
+
+class FailedInContactingTestPartner(Exception):
+    pass
+
+
+class TestPartnerClient(object):
+    # This is valid for v.6 for the parters API
+
+    def __init__(self, config):
+        # TODO Figure out how to configure this, given Claritys normal setup
+        super().__init__()
+        self._url = config.get("test_partner_url")
+        self._user = config.get("test_partner_user")
+        self._password = config.get("test_partner_password")
+
+    # TODO decide if we want to retry or not. It depends a little bit on the
+    #      rest of the scenario. This could potentially take us into a long chain
+    #      of retries. That would mean things take a long time before failing.
+    #      Don't know what is best here.
+    @retry((ConnectionError, Timeout), tries=3, delay=2, backoff=2)
+    def send_single_sample_result(self, test_partner_sample_info):
+
+        if not isinstance(TestPartnerSampleInformation):
+            raise AssertionError(
+                "Expected type of test_partner_sample_info is TestPartnerSampleInformation")
+
+        parameters = test_partner_sample_info.get_as_dict().update(
+            {"user": self._user, "password": self._password})
+        headers = {"Content-Type": "application/x-www-form-urlencoded"}
+
+        response = requests.post(
+            self._url, header=headers, data=parameters)
+
+        if not response.status_code == 200:
+            raise FailedInContactingTestPartner("Did not get a 200 response from test partner. " +
+                                                "Response status code was: {}" +
+                                                "and response text: {}".format(response.status_code, response.text))
+
+        # Example of successful API call response
+        # success
+        # referrral_code=1234567890
+        # covid19_result=failed
+        # arrival_date=20140310
+        # test_date=20140310
+        # result_date=20140310
+        # comment=ett testmeddelande
+        response_text_parts = response.text.split("\n")
+        if response_text_parts[0] == "success":
```

Comment:
Right. I think at least that we would want the error codes to be reported to the end user and logged, particularly if we're going to run this on a button click from Clarity.

I think it would be fine to just raise an execption in the case of the error codes and return silently if not.

Reply:
I'll come back to this tomorrow and have it raise it was not succesful.